fair/tests/unittests/test_fair.py

```diff
@@ -0,0 +1,96 @@
+
+F=fair.Fair(predictor,'sex',test_data)
+
+
+#Maximize accuracy while enforcing weak equalized odds,
+#that the difference in recall between groups is less than 2.5% 
+
+F.maximize(overall.accuracy, group_metric(overall.recall,'max_diff'),0.025,'<')
+
+#Evaluate the change in fairness (recall difference corresponds to EO)
+measures = F.evaluate_fairness()
+
+assert measures['Original Classifier']['Recall Difference (equalized odds)']>0.025
+
+assert measures['Updated Classifier']['Recall Difference (equalized odds)']<0.025
+
+
+#set up new fair class using 'race' as the protected group and evaluate on test data
+F=fair.Fair(predictor,'race',test_data)
+
+full_group_metrics=F.evaluate_groups()
+
+#drop the two smallest groups and reevaluate
+new_test=test_data[~test_data['race'].isin([' Other',' Asian-Pac-Islander',])] #drop other
+F=fair.Fair(predictor,'race',new_test)
+partial_group_metrics=F.evaluate_groups()
+
+#Check that metrics computed over a subset of the data is consistent with metrics over all data
+for group in (' White', ' Black', ' Amer-Indian-Eskimo'):
+    assert all(full_group_metrics.loc[group]==partial_group_metrics.loc[group])
+
+assert all(full_group_metrics.loc['Maximum difference']>=partial_group_metrics.loc['Maximum difference'])
+
+
+#Enforce the 4/5 rule that the max ratio between the proportion of positive decisions is less than 1.25
+F.maximize(overall.accuracy,group_metric(overall.expected_prediction,'max_ratio'),1.25,'<')
+#Evaluate the change in fairness (recall difference corresponds to EO)
+measures = F.evaluate_fairness() 
+
+assert measures['Original Classifier']['Disparate Impact']>1.25
+
+assert measures['Updated Classifier']['Disparate Impact']<1.25
+
+#Enforce that every group has a recall over 0.5
+F.maximize(overall.accuracy, group_metric(overall.recall, 'min'),0.5,'>')
+group_metrics=F.evaluate_groups()
+assert all(group_metrics['Recall'][:-1]>0.5)
+
+#train two new classifiers one to predict class without using sex and one to predict sex without using class
+predictor,protected=fair.learners.infered_attribute_builder(train_data,'class','sex')
+#Build fair object using this and evaluate fairness n.b. classifier accuracy decreases due to lack of access to the protected attribute, but otherwise code is doing the same thing 
+F=fair.Fair(predictor,'sex',test_data,infered_groups=protected)
+
+#Enforce that the new classifier will satisfy equalised odds (recall difference between protected attributes of less than 2.5%) despite not using sex at run-time  
+F.maximize(overall.accuracy,group_metric(overall.recall,'max_diff'),0.025,'<')
+
+measures = F.evaluate_fairness()
+
+assert measures['Original Classifier']['Recall Difference (equalized odds)']>0.025
+
+assert measures['Updated Classifier']['Recall Difference (equalized odds)']<0.025
+
+#Prove that sex isn't being used by dropping it and reevaluating.
+
+new_data=test_data.drop('sex',axis=1,inplace=False)
+F.evaluate_groups(new_data,test_data['sex'])
```

Comment:
Missing test of predict and predict_proba, we never show that the code actually works to make the model return new predictions.

Reply:
done